hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java

```diff
@@ -410,61 +416,183 @@ public boolean rename(final Path src, final Path dst) throws IOException {
     Path qualifiedSrcPath = makeQualified(src);
     Path qualifiedDstPath = makeQualified(dst);
 
+    Path nestedDstParent = dst.getParent();
+    if (getAbfsStore().getAbfsConfiguration().getPrefixMode()
+        == PrefixMode.BLOB) {
+      /*
+       * For blob endpoint with non-HNS account, client has to ensure that destination
+       * is not a sub-directory of source.
+       */
+      LOG.debug("Check if the destination is subDirectory");
+      while (nestedDstParent != null) {
+        if (makeQualified(nestedDstParent).equals(qualifiedSrcPath)) {
+          //testRenameChildDirForbidden.
+          LOG.info("Rename src: {} dst: {} failed as dst is subDir of src",
+              qualifiedSrcPath, qualifiedDstPath);
+          return false;
+        }
+        nestedDstParent = nestedDstParent.getParent();
+      }
+    }
+
+
     TracingContext tracingContext = new TracingContext(clientCorrelationId,
         fileSystemId, FSOperationType.RENAME, true, tracingHeaderFormat,
         listener);
     // rename under same folder;
-    if(makeQualified(parentFolder).equals(qualifiedDstPath)) {
+    if (makeQualified(parentFolder).equals(qualifiedDstPath)) {
       return tryGetFileStatus(qualifiedSrcPath, tracingContext) != null;
     }
 
-    FileStatus dstFileStatus = null;
     if (qualifiedSrcPath.equals(qualifiedDstPath)) {
       // rename to itself
       // - if it doesn't exist, return false
       // - if it is file, return true
       // - if it is dir, return false.
-      dstFileStatus = tryGetFileStatus(qualifiedDstPath, tracingContext);
-      if (dstFileStatus == null) {
+      final AtomicBoolean isDstDirectory = new AtomicBoolean();
+      final AtomicBoolean isDstExists = new AtomicBoolean();
+
+      getPathInformation(qualifiedDstPath, tracingContext, isDstDirectory,
+          isDstExists);
+      if (!isDstExists.get()) {
         return false;
       }
-      return dstFileStatus.isDirectory() ? false : true;
+      return isDstDirectory.get() ? false : true;
     }
 
+    final AtomicBoolean isDstDirectory = new AtomicBoolean();
```

Comment:
repetitive variable usage, previous variable can be defined outside the loop.

Reply:
It is not a a problem. Since, scopes are different. But can be used the same object. Thing is even now,
its either going to get init in the loop or out. Wont be init at both place.
For the comment, taking it.

Reply:
Using different variable names in this iteration, since, usage is different.